lib/CL/pocl_cl.h

```diff
@@ -987,6 +987,10 @@ typedef struct pocl_global_mem_t {
 
 #define NUM_OPENCL_IMAGE_TYPES 6
 
+/* typedef for intrinsics replacement function. TODO this is a hack,
```

Comment:
I wonder should we call this a temporary hack or just accept it as a new feature that could be useful for other targets for converting intrinsics to lib functions without updating the LLVM backend. Logically, these cross-AS-memcpys are for OpenCL and PoCL is an OpenCL implementation so it should provide the implementations (for the targets). I'd just make it an "official" device hook for now and document it properly.

Reply:
> accept it as a new feature that could be useful for other targets

It might be useful, but it's still a hack to fix deficiencies in the backend(s). And it is not maintenance free, because right now memcpy cannot be implemented in C source (because of the i1 argument) only in '.ll' and the format of that is LLVM version dependent.

> Logically, these cross-AS-memcpys are for OpenCL and PoCL is an OpenCL implementation so it should provide the implementations (for the targets).

From what i read, Clang and GCC both can emit calls to functions like memcpy independently of the source language, and it is the responsibility of the runtime to provide the implementations for memset, memcpy, memmove and memcmp. And if these functions are to be efficient, they need target specific knowledge. So IMO this does not belong into PoCL and should not be made "official".

Reply:
OK, let's generalize when there's another target needing this.

> From what i read, Clang and GCC both can emit calls to functions like memcpy independently of the source language, and it is the responsibility of the runtime to provide the implementations for memset, memcpy, memmove and memcmp. And if these functions are to be efficient, they need target specific knowledge. So IMO this does not belong into PoCL and should not be made "official".

For "target specific implementations" we have the target-specific built-in library override mechanism. Sure, we can push it to a separate builtin library, but it doesn't fit to libc due to libc being a C library (address space 0 memcpys etc. belong there), then the question is what is the correct library/place for these if not in an OpenCL (compiler) implementation of which multi-AS compiler is producing them? It's not an entirely clear answer to me. I wonder where the other targets put them or just expand in the backend to IR? CPUs just flatten out the AS so they can use the normal memcpy.

Reply:
> I wonder where the other targets put them or just expand in the backend to IR?

AFAICT, you can't put them anywhere. LLVM refuses to compile an .ll file which actually defines (not just declares) an intrinsic. So it seems to me the intention is that the backend is supposed to expand the intrinsic.